controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
 	}
 
-	vrgMWName := fmt.Sprintf(ManifestWorkNameFormat, subscription.Name, subscription.Namespace, MWTypeVRG)
-
-	err = a.deleteExistingManifestWork(subscription, vrgMWName)
-	if err != nil {
-		a.log.Error(err, "Failed to delete existing VRG manifestwork for subscription", "name", subscription.Name)
+	if homeCluster == "" && len(a.avrPlacementRule.Status.Decisions) != 0 {
+		homeCluster = a.avrPlacementRule.Status.Decisions[0].ClusterName
+		homeClusterNamespace = a.avrPlacementRule.Status.Decisions[0].ClusterNamespace
+	}
 
-		return !unpause
+	if homeCluster == "" {
+		return !done, fmt.Errorf("PreferredCluster not set and unable to find home cluster in AVRPlacementRule (%s)",
+			a.avrPlacementRule.Name)
 	}
 
-	err = a.restorePVFromBackup(subscription, newHomeCluster)
+	// We have a home cluster
+	err := a.updateUserPlRuleAndCreateVRGMW(homeCluster, homeClusterNamespace)
 	if err != nil {
-		a.log.Error(err, "Failed to restore PVs from backup for subscription", "name", subscription.Name)
+		return !done, err
+	}
 
-		return !unpause
+	// All good, update the preferred decision and state
+	if len(a.userPlacementRule.Status.Decisions) > 0 {
+		a.instance.Status.PreferredDecision = a.userPlacementRule.Status.Decisions[0]
 	}
 
-	a.moveToNextDRState(subscription.Name)
+	a.advanceToNextDRState()
+
+	a.log.Info(fmt.Sprintf("AVR (%+v)", a.instance))
 
-	return unpause
+	return done, nil
 }
 
-func (a *AVRInstance) vrgManifestWorkAlreadyExists(
-	subscription *subv1.Subscription) (bool, error) {
-	if a.instance.Status.Decisions == nil {
-		return false, nil
+//
+// runFailover:
+// 1. If failoverCluster empty, then fail it and we are done
+// 2. If already failed over, then clean up and we are done
+// 3. Set VRG in the preferredCluster to secondary
+// 4. Restore PV to failoverCluster
+// 5. Update UserPlacementRule decision to failoverCluster
+// 6. Create VRG for the failoverCluster as Primary
+// 7. Update AVR status
+// 8. Delete VRG MW from preferredCluster once the VRG state has changed to Secondary
+//
+func (a *AVRInstance) runFailover() (bool, error) {
+	a.log.Info("Entering runFailover", "state", a.instance.Status.LastKnownDRState)
+
+	const done = true
+
+	// We are done if empty
+	if a.instance.Spec.FailoverCluster == "" {
+		return done, fmt.Errorf("failover cluster not set. FailoverCluster is a mandatory field")
 	}
 
-	if d, found := a.instance.Status.Decisions[subscription.Name]; found {
-		// Skip this subscription if a manifestwork already exist for it
-		mwName := BuildManifestWorkName(subscription.Name, subscription.Namespace, MWTypeVRG)
+	newHomeCluster := a.instance.Spec.FailoverCluster
 
-		mw, err := a.findManifestWork(mwName, d.HomeCluster)
-		if err != nil {
-			a.log.Error(err, "findManifestWork()", "name", subscription.Name)
+	// We are done if we have already failed over
+	if len(a.userPlacementRule.Status.Decisions) > 0 {
+		if a.instance.Spec.FailoverCluster == a.userPlacementRule.Status.Decisions[0].ClusterName {
```

Comment:
If runPlacementTask->updateUserPlRuleAndCreateVRGMW manages to update placement rule and fails or leaves VRG creation incomplete, then we will never set VRG as primary on the current cluster in the next iteration?

Should updating the userPl be the last operation hence, denoting we are done placing the workload and may have cleanup actions pending?